File: src/windows/AmbientContextMcp.Core/Models/TransmissionUiSettingsMerge.cs
namespace AmbientContextMcp.Core.Models;

public static class TransmissionUiSettingsMerge
{
    public static Dictionary<string, bool> MergeOverrides(
        IReadOnlyDictionary<string, bool> existingOverrides,
        IReadOnlyList<TransmissionUiOptionDefinition> options,
        IReadOnlySet<string> enabledOptionIds)
    {
        var overrides = new Dictionary<string, bool>(
            existingOverrides,
            StringComparer.OrdinalIgnoreCase);

        var managedPaths = options
            .SelectMany(option => option.LinkedPaths)
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToHashSet(StringComparer.OrdinalIgnoreCase);
        var allowedPaths = options
            .Where(option => enabledOptionIds.Contains(option.Id))
            .SelectMany(option => option.LinkedPaths)
            .ToHashSet(StringComparer.OrdinalIgnoreCase);

        foreach (var path in managedPaths)
        {
            if (allowedPaths.Contains(path))
            {
                overrides[path] = true;
            }
            else
            {
                overrides.Remove(path);
            }
        }

        return overrides;
    }

    public static bool IsOptionEnabled(
        string primaryPath,
        IReadOnlyDictionary<string, bool> overrides)
    {
        return overrides.TryGetValue(primaryPath, out var allowed) && allowed;
    }

    public static bool IsOptionEnabled(
        TransmissionUiOptionDefinition option,
        IReadOnlyDictionary<string, bool> overrides)
    {
        return IsOptionEnabled(option.PrimaryPath, overrides);
    }
}

File: src/macos/Sources/AmbientContextCore/Models/TransmissionUiSettingsMerge.swift
import Foundation

public enum TransmissionUiSettingsMerge {
    /// UI で ON のオプションが持つ LinkedPaths の和集合を override に反映する。
    /// UI が管理していない path (手編集の legacy override 等) はそのまま残す。
    public static func mergeOverrides(
        existingOverrides: CaseInsensitiveDictionary<Bool>,
        options: [TransmissionUiOptionDefinition],
        enabledOptionIds: Set<String>
    ) -> CaseInsensitiveDictionary<Bool> {
        var overrides = existingOverrides

        var managedPaths: [String] = []
        var seen = Set<String>()
        for option in options {
            for path in option.linkedPaths where seen.insert(path.lowercased()).inserted {
                managedPaths.append(path)
            }
        }

        let normalizedEnabledIds = Set(enabledOptionIds.map { $0.lowercased() })
        var allowedPaths = Set<String>()
        for option in options where normalizedEnabledIds.contains(option.id.lowercased()) {
            for path in option.linkedPaths {
                allowedPaths.insert(path.lowercased())
            }
        }

        for path in managedPaths {
            if allowedPaths.contains(path.lowercased()) {
                overrides[path] = true
            } else {
                overrides.removeValue(forKey: path)
            }
        }

        return overrides
    }

    public static func isOptionEnabled(
        primaryPath: String,
        overrides: CaseInsensitiveDictionary<Bool>
    ) -> Bool {
        overrides[primaryPath] == true
    }

    public static func isOptionEnabled(
        option: TransmissionUiOptionDefinition,
        overrides: CaseInsensitiveDictionary<Bool>
    ) -> Bool {
        isOptionEnabled(primaryPath: option.primaryPath, overrides: overrides)
    }
}
